File: clojure/src/advent/day5.clj
(ns advent.day5
	(:require [clojure.string :refer [split]]))


(defn nice? [s]
	(and 
		(re-find #"[aeiou].*[aeiou].*[aeiou]" s)
		(re-find #"(.)\1" s)
		(not (re-find #"ab|cd|pq|xy" s))))

(defn nicer? [s]
	(and
		(re-find #"(.).\1" s)
		(re-find #"(..).*\1" s)))

(defn count-nice [pred s]
	(count
		(filter pred
			(split s #"\n"))))


(defn day5 []
	(let [s (slurp "../data/day5.txt")]
		(do
			(println "****** Day 5 *******")
			(println "Santa has" (count-nice nice? s) "nice strings in his file.")
			(println "Santa has" (count-nice nicer? s) "nicer strings in his file.")
			(println ))))

File: js/lib/day5.js
import fs from 'fs';

export default function day5() {
  var data = fs.readFileSync('../data/day5.txt', 'utf-8');
  console.log('**** Day 5 *****');
  console.log('Santa has ', count(isNice, data), 'nice strings in his file.');
  console.log('Santa has ', count(isNicer, data), 'nicer strings in his file.');
  console.log();
}

function count(pred, data) {
  return data.split('\n').filter(pred).length;
}

export function isNice(s) {
  return /(.)\1/.test(s) &&
         /[aeiou].*[aeiou].*[aeiou]/.test(s) &&
         !/ab|cd|pq|xy/.test(s);
}

export function isNicer(s) {
  return /(.).\1/.test(s) &&
         /(..).*\1/.test(s);
}

